File: src/test/java/de/espend/idea/php/annotation/tests/doctrine/action/DoctrineAddRepositoryGenerateActionTest.java
package de.espend.idea.php.annotation.tests.doctrine.action;

import com.jetbrains.php.lang.PhpFileType;
import de.espend.idea.php.annotation.tests.AnnotationLightCodeInsightFixtureTestCase;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see de.espend.idea.php.annotation.doctrine.action.DoctrineAddRepositoryGenerateAction
 */
public class DoctrineAddRepositoryGenerateActionTest extends AnnotationLightCodeInsightFixtureTestCase {
    public void setUp() throws Exception {
        super.setUp();
        myFixture.copyFileToProject("DoctrineAddRepositoryGenerateAction.php");
    }

    public String getTestDataPath() {
        return "src/test/java/de/espend/idea/php/annotation/tests/doctrine/action/fixtures";
    }

    public void testGenerationForAnnotation() {
        myFixture.configureByText(PhpFileType.INSTANCE, "<?php\n" +
            "" +
            "namespace App\\Entity;" +
            "\n" +
            "use Doctrine\\ORM\\Mapping as ORM;\n" +
            "\n" +
            "/**\n" +
            " * @ORM\\Entity()\n" +
            " */\n" +
            "class User\n" +
            "{" +
            "<caret>\n" +
            "}"
        );

        myFixture.performEditorAction("PhpAnnotation.Doctrine.Orm.DoctrineAddRepositoryGenerateAction");

        myFixture.checkResult("<?php\n" +
            "namespace App\\Entity;\n" +
            "use App\\Entity\\Repository\\UserRepository;\n" +
            "use Doctrine\\ORM\\Mapping as ORM;\n" +
            "\n" +
            "/**\n" +
            " * @ORM\\Entity(repositoryClass=UserRepository::class)\n" +
            " */\n" +
            "class User\n" +
            "{\n" +
            "}"
        );
    }

    public void testGenerationForAnnotationInsideDocCommentScope() {
        myFixture.configureByText(PhpFileType.INSTANCE, "<?php\n" +
            "" +
            "namespace App\\Entity;" +
            "\n" +
            "use Doctrine\\ORM\\Mapping as ORM;\n" +
            "\n" +
            "/**\n" +
            " * @ORM\\Ent<caret>ity()\n" +
            " */\n" +
            "class User\n" +
            "{" +
            "\n" +
            "}"
        );

        myFixture.performEditorAction("PhpAnnotation.Doctrine.Orm.DoctrineAddRepositoryGenerateAction");

        myFixture.checkResult("<?php\n" +
            "namespace App\\Entity;\n" +
            "use App\\Entity\\Repository\\UserRepository;\n" +
            "use Doctrine\\ORM\\Mapping as ORM;\n" +
            "\n" +
            "/**\n" +
            " * @ORM\\Entity(repositoryClass=UserRepository::class)\n" +
            " */\n" +
            "class User\n" +
            "{\n" +
            "}"
        );
    }
}

File: src/test/kotlin/de/espend/idea/php/annotation/tests/doctrine/action/DoctrineAddRepositoryGenerateActionTest.kt
package de.espend.idea.php.annotation.tests.doctrine.action

import com.jetbrains.php.lang.PhpFileType
import de.espend.idea.php.annotation.tests.AnnotationLightCodeInsightFixtureTestCase

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see de.espend.idea.php.annotation.doctrine.action.DoctrineAddRepositoryGenerateAction
 */
class DoctrineAddRepositoryGenerateActionTest : AnnotationLightCodeInsightFixtureTestCase() {
    public override fun setUp() {
        super.setUp()
        myFixture.copyFileToProject("DoctrineAddRepositoryGenerateAction.php")
    }

    override fun getTestDataPath(): String {
        return "src/test/kotlin/de/espend/idea/php/annotation/tests/doctrine/action/fixtures"
    }

    fun testGenerationForAnnotation() {
        myFixture.configureByText(
            PhpFileType.INSTANCE,
            """
                <?php
                namespace App\Entity;
                use Doctrine\ORM\Mapping as ORM;

                /**
                 * @ORM\Entity()
                 */
                class User
                {<caret>
                }
            """.trimIndent(),
        )

        myFixture.performEditorAction("PhpAnnotation.Doctrine.Orm.DoctrineAddRepositoryGenerateAction")

        myFixture.checkResult(
            """
                <?php
                namespace App\Entity;
                use App\Entity\Repository\UserRepository;
                use Doctrine\ORM\Mapping as ORM;

                /**
                 * @ORM\Entity(repositoryClass=UserRepository::class)
                 */
                class User
                {
                }
            """.trimIndent(),
        )
    }

    fun testGenerationForAnnotationInsideDocCommentScope() {
        myFixture.configureByText(
            PhpFileType.INSTANCE,
            """
                <?php
                namespace App\Entity;
                use Doctrine\ORM\Mapping as ORM;

                /**
                 * @ORM\Ent<caret>ity()
                 */
                class User
                {
                }
            """.trimIndent(),
        )

        myFixture.performEditorAction("PhpAnnotation.Doctrine.Orm.DoctrineAddRepositoryGenerateAction")

        myFixture.checkResult(
            """
                <?php
                namespace App\Entity;
                use App\Entity\Repository\UserRepository;
                use Doctrine\ORM\Mapping as ORM;

                /**
                 * @ORM\Entity(repositoryClass=UserRepository::class)
                 */
                class User
                {
                }
            """.trimIndent(),
        )
    }
}
